File: 04/04.clj
(ns dec04)
(require '[clojure.string :as string])

(defn sorted-string? [s]
  (= (sort s) s))

(defn has-duplicate? [s]
  (->> (set s)
       (count)
       (> 6)))

(defn has-two-of-something? [s]
  (contains?
   (->> (frequencies s)
        (vals)
        (set))
   2))

(defn digit-split [n]
  (string/split (str n) #""))

(defn is-valid-first [digits]
  (and (sorted-string? digits)
       (has-duplicate? digits)))

(defn is-valid-second [digits]
  (and (sorted-string? digits)
       (has-two-of-something? digits)))

(defn valid-in-range [valid from to]
  (->> (range from to)
       (filter #(valid (digit-split %)))
       (count)))

(println "Part 1:" (valid-in-range is-valid-first 240920 789957))
(println "Part 2:" (valid-in-range is-valid-second 240920 789957))

File: 2019/04/04.py
def is_valid_first(n):
    digits = [int(j) for j in str(n)]
    if digits != sorted(digits):
        return False

    for digit in digits:
        if digits.count(digit) > 1:
            return True


def is_valid_second(n):
    digits = [int(j) for j in str(n)]
    if digits != sorted(digits):
        return False

    for digit in digits:
        if digits.count(digit) == 2:
            return True


valid_first = 0
valid_second = 0
for n in range(240920, 789957):
    if is_valid_first(n):
        valid_first += 1
    if is_valid_second(n):
        valid_second += 1

print("Part 1:", valid_first)
print("Part 2:", valid_second)
